arbnode/transaction_streamer.go

```diff
@@ -1558,6 +1542,53 @@ func (s *TransactionStreamer) HasNotSubmitted(pos arbutil.MessageIndex) (bool, e
 	return true, nil
 }
 
+/**
+* If the transaction fails to fetch, we will retry for a while before resubmitting
+* the transaction to the Espresso node.
+ */
+func (s *TransactionStreamer) tryResubmittingEspressoTransactions(ctx context.Context, firstSubmitted SubmittedEspressoTx, submittedTxHash *tagged_base64.TaggedBase64, err error) error {
+	if s.lastSubmitFailureAt == nil {
+		now := time.Now()
+		s.lastSubmitFailureAt = &now
+		return fmt.Errorf("trying to resubmit transaction (hash: %s): %w, will retry", submittedTxHash.String(), err)
+	}
+
+	duration := time.Since(*s.lastSubmitFailureAt)
+	if duration < s.resubmitEspressoTxDeadline {
+		return fmt.Errorf("not enough time to attempt resubmission of transaction (hash: %s): %w, will retry again", submittedTxHash.String(), err)
+	}
+	txHash, err := s.resubmitEspressoTransactions(ctx, firstSubmitted)
+	if err != nil {
+		return fmt.Errorf("failed to resubmit transaction (hash: %s): %w", submittedTxHash.String(), err)
+	}
+	// The first submitted tx should be updated to the resubmitted tx
+	submittedTxns, err := s.getEspressoSubmittedTxns()
+	if err != nil {
+		return fmt.Errorf("trying to resubmit transaction: failed to get the submitted txns: %w", err)
+	}
+	tx := SubmittedEspressoTx{
+		Hash:    txHash.String(),
+		Pos:     firstSubmitted.Pos,
+		Payload: firstSubmitted.Payload,
+	}
+	submittedTxns[0] = tx
+	batch := s.db.NewBatch()
+	err = s.setEspressoSubmittedTxns(batch, submittedTxns)
+	if err != nil {
+		return fmt.Errorf("trying to resubmit transaction: failed to set the submitted txns: %w", err)
+	}
+	err = batch.Write()
+	if err != nil {
+		return fmt.Errorf("failed to write to db: %w", err)
+	}
```

Comment:
It feels like we shouldn't have to do this. Since we're just resubmitting a transaction, the hash and all the contents should be the same. You can think of it like we're submitting a transaction with hash `H` and polling persistently until we see that it gets finalized. The resubmitting is just a detail that is sometimes needed to ensure a transaction -- *with hash `H`* -- does eventually get finalized. So if this is needed to make it work in the test I want to understand why

Reply:
sorry i had ethereum brain where resubmission would usually happen due to a change in gas price which effectively changes the txn hash as well
if espresso deterministically gets the txn hash based on the payload alone it should be the same, ill test it out

Reply:
verified you're correct. will remove this.